worker/runtime/backend.go

```diff
@@ -28,6 +29,10 @@ type GardenBackend struct {
 	network       Network
 	rootfsManager RootfsManager
 	userNamespace UserNamespace
+
+	maxContainers  int
+	requestTimeout time.Duration
+	createLock     *semaphore.Weighted
```

Comment:
FWIW, when I suggested using semaphores here https://github.com/concourse/concourse/issues/5226#issuecomment-642094627, I didn't have a mutex in mind. I was thinking more along the lines of:

* On startup, check the number of containers by querying `containerd`
* Initialize a semaphore with value `MaxContainerLimit - NumContainersOnStartup`
* Whenever you create a container, acquire the semaphore
* Whenever you destroy a container, release the semaphore

If it's not too much of a performance burden querying `containerd` for the list of containers whenever you start a container, the approach you took is definitely safer, since we don't have to worry about containers going away without us releasing the semaphore. Figured I'd bring up the alternative approach, anyway

Reply:
We were thinking about that approach and @xtreme-sameer-vohra brought up the concern that there could be issues if the worker binary ever gets out of sync with containerd. In order to rectify that we would have to periodically query number of containers and update our count. But we would not be able to use `semaphore.Weighted` for that as it does not expose the count / weight, and if we implement our own, it eventually leads down a rabbit hole of details.

The only reason we used semaphores was that we wanted to put a timeout on lock acquisition and the semaphore is the only synchronization primitive that took a `context.Context`.

Reply:
Daniel covered the reasons for our approach - primarily making sure we're in sync
This was also the approach that the current [Guardian implementation uses](https://github.com/cloudfoundry/guardian/blob/6b021168907b2f2ae25cf54acae8b454b430332f/gardener/gardener.go#L230-L241) - albeit without a lock.